### DISCUSSION TITLE
Welcome to BLE HID Extension Discussions!

### DISCUSSION THREAD

**Discussion by bsiever** [bsiever/microbit-pxt-blehid#16] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by MrG4Real** [comment]
Hi! I'm a middle school teacher who uses Micro:bit v2 in the classroom. I'm interested in using this extension, but I don't know how to get started. My students use Makecode to programme our Micro:bits. Any assistance would be appreciated.

It would make a big difference to my students if we could map just a few micro:bit inputs to keyboard or mouse.

Thanks,
Mr G

**Comment by bsiever** [comment]
Hi @MrG4Real :  Do you have any thoughts for activities you may use it for?  If you haven't already seen it, here's a video that shows some of the things it can do: https://www.youtube.com/watch?v=n4J5GN72N_4

**Reply by MrG4Real** [reply]
Hi! 

Thanks for the quick reply! Intially I simply want students to utilise buttons (or tilt or compass bearings) to act as mouse/keys for whatever digital outcome they are working on. For example, currently I have students making a game where a double-hulled canoe navigates the Pacific Ocean. I would love for a physical model of the canoe to have a micro-bit inside, so it could respond to tilt.

My initial hurdle is I have no idea how to bring up the coding blocks shown in that video:
- Bluetooth Keyboard Service
- Send keys

I guess there's an extension I need to add, but don't know where to begin.

Thanks,
Mr G

**Reply by bsiever** [reply]
Ah!  Sure: 
1. Create a new project
2. Click the button for "Advanced" blocks
3.  Click "+ Extension"
4. Paste `https://github.com/bsiever/microbit-pxt-blehid` in the box that says "Search or enter project URL"
5. Hit "Enter" 
6. Then click on the tile for the extension
7. You'll be prompted to confirm you want to remove the Radio blocks (you can either do Bluetooth or Radio, not both).  Confirm you do.

Or you can open this example project: https://makecode.microbit.org/_Fusc23d9R9tc

The extension is being considered to become an official extension, which would make it easier to add.

Let me know if that helps!

**Reply by MrG4Real** [reply]
Thanks. Steps 1 - 4 gave me new blocks to play with, but I can't see any content in Makecode or in the example project. It only has the "on start" and  "forever" blocks.

**Reply by bsiever** [reply]
Oh --- I didn't include any content, I just thought an existing project may be an easier way to get the blocks.  If you click on the "Edit" button you could customize it / add blocks. 

By the way, if you're interested in seeing how it can be used for Accessibility Devices, I'll help with a workshop in a few weeks:  
<blockquote class="twitter-tweet"><p lang="en" dir="ltr">Learn how to build an assistive switch w/o soldering or 3D printing with <a href="https://twitter.com/BillSiever?ref_src=twsrc%5Etfw">@BillSiever</a> and chapter leader <a href="https://twitter.com/loretod101?ref_src=twsrc%5Etfw">@loretod101</a>.<br><br>📍 Virtual<br>📅 Wed, April 13 at 5pm EST<br>🛠️ Beginner friendly<br><br>Register: <a href="https://t.co/KYyWjXKJrM">https://t.co/KYyWjXKJrM</a></p>&mdash; Makers Making Change (@MakerMakeChange) <a href="https://twitter.com/MakerMakeChange/status/1509941169521934336?ref_src=twsrc%5Etfw">April 1, 2022</a></blockquote>

**Reply by MrG4Real** [reply]
Thanks so much for your help. It was exactly what I needed, and my students got something out of it to enhance their projects.

**Reply by bsiever** [reply]
That's good to hear!

By the way, it's an official extension now, so it's a little easier to add to projects:
1. Create/Open a project
2. Click the gear icon in the upper right
3. Select "Extensions" 
4. Type "BLE HID" in the "Search or enter project URL..."  and hit enter.
5. Click on the microbit-pxt-blehid tile. 
6. You'll be prompted to remove the "Radio" blocks. Select the "Remote extension(s) and add microbit-pxt-blehid" button.

**Comment by NickAskewGH** [comment]
Hi, I have been looking at various small computing platforms with a view to emulating BLE HIDs, specifically a mouse and keyboard for automation purposes and came across your amazing makecode extension for the microbit.

Now armed with a microbit I have been creating python projects outside of the makecode environment (I use PyCharm with a micropython extension that is microbit aware and uses uflash to push the built hex file to the device) and can access the buttons, sensors and uart interface which is provided care of "import microbit import *"

However, I cannot understand how to link your library into it (outside of the makecode environment).  If you are aware this is possible would you be able to point me in the right direction as a starting point please?

Many thanks for any assistance you could offer!
Nick

**Reply by bsiever** [reply]
Hi @NickAskewGH, 

The extension is specific to MakeCode, but can probably be used from MakeCode's version of Python.  Much of my work is focused on lightweight activities for kids from ages 8-12, so I mainly develop MakeCode extensions for block based development.  

There are two major flavors of Python that can run on the Micro:bit.  MakeCode's version is actually translated to machine code (traditional compilation) before download and is only supported from the MakeCode tools.  Other editors ,like PyCharm and the Microbit online Python editor, usually use a MicroPython version running on the micro:bit. I think MicroPython uses a Python bytecode based approach.  See https://microbit-micropython.readthedocs.io/en/v2-docs/ and https://github.com/micropython/micropython .  

MicroPython has somewhat limited Bluetooth support (see: https://microbit-micropython.readthedocs.io/en/v2-docs/ble.html).  My work is built on the CODAL micro:bit runtime (https://github.com/lancaster-university/codal-microbit-v2) and it looks like there is MicroPython support for CODAL (https://github.com/microbit-foundation/micropython-microbit-v2), so it _might_ be possible to port it to work with MicroPython. (I don't have time to look into this or do it until at least May).

**Reply by NickAskewGH** [reply]
Hi @bsiever,

Thank you for responding so quickly! 

I really appreciate all the links you've taken the time to provide, I'll take a good look at the CODAL runtime.

Best wishes
Nick

**Comment by RocioCabral** [comment]
Hi! I find this extension really interesting, but my microbit is V1. Do you know if there are any similar extensions, that could work for V1 microbits?

Thanks,
Rocío

**Reply by bsiever** [reply]
Hi @RocioCabral , 

There is an older extension that is no longer supported with the current version of MakeCode that did work for v1.  I think it can still be used if you use an older version of MakeCode though.   Please see this discussion: https://forum.makecode.com/t/anyone-ever-used-microbit-as-a-controller/9959/8 

Unfortunately, I can't provide much more guidance.
